### DISCUSSION TITLE
Update-2023-04-24 Molecules

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#215] [Announcements]
Just for fun I wrote some code to parse and render molecular structures. It can read in `.xyz` and `.mol` files.

The code is in `src/plugins/molecule.lisp` with demos in `test/demo-misc.lisp`.

I got my start in 3D graphics back in the 80's when I got a summer job at Xerox Research Centre of Canada (XRCC) to write a package for rendering molecular graphics. I wrote my first ray tracer there (in Fortran running on a DEC VAX 11/750), so this was a bit of a trip down memory lane:

https://medium.com/@kaveh808/8-summer-at-xerox-in-24-bit-color-74ad465a07be

Below is a caffeine molecule.

![Screen Shot 2023-04-14 at 1 17 39 PM](https://user-images.githubusercontent.com/103860793/234149561-6313b9d6-9d2f-4ea2-ad9a-4e63bfda0b76.png)

**Comment by aykaramba** [comment]
Awesome!

**Comment by kayomarz** [comment]
This is pretty neat!

I have a question:

On evaluating the [demo code](https://github.com/kaveh808/kons-9/blob/main/test/demo-misc.lisp#L109-L146), the bond lines did not appear as it does in your screenshot. Just out of curiosity was wondering if there's any size of the  spheres/bond lines that need to be set?

Here's what I see:

![molecules-without-bond-lines-screenshot](https://user-images.githubusercontent.com/140297/234373969-d143a8a9-20c8-4efe-a084-f1cb49029954.png)

**Reply by kaveh808** [reply]
The XYZ file format does not contain bond data. The MOL file format does. If you look at the second demo for MOL you will see :show-bonds? set to t.

**Reply by kayomarz** [reply]
Ah, I missed seeing that last expression with the :show-bonds. Thanks:)

**Comment by rigidus** [comment]
I would like to modify the engine to be able (for example) to select individual atoms with a mouse click (and possibly display a dialog box next to it with help from the periodic table of elements). 

I see that there are all the prerequisites for this - dialog boxes, ability to move elements, etc., but I'd like to understand how I can match the coordinates of the mouse pointer to the area of the displayed element

Where should I start?

**Reply by kaveh808** [reply]
This is a significant undertaking which would require a number of code enhancements.

Off the top of my head:

1) Detect mouse clicks in the viewport (in kons-9/src/graphics/glfw/glfw-gui.lisp) and map the pixel to a view 3D ray.

2) Implement ray-polyhedron intersection.

3) Detect polyhedron object clicked on. (You can Google "3d OpenGL picking" for this)

4) Enhance the shape class to handle custom annotations (e.g. atom type in the case of a molecule).

5) Allow for the display of the annotation data as a panel in the viewport.

Hope this helps.
